From 679bfeceb96284d81d763950fc61e2826f282dfc Mon Sep 17 00:00:00 2001
From: Yuanyuan Liu <yuanliu@codeaurora.org>
Date: Wed, 10 May 2017 12:05:11 -0700
Subject: [PATCH 2/4] wifi: Check whether dev node is accessible or not before
 opening

wifi framework tries to open dev node WIFI_DRIVER_STATE_CTRL_PARAM
as part of initialization/turn on. But this dev node may not be
present/not accessible by then. So check for the
existence/accessibility of dev node by retrying few times to
avoid possible race condition between creation and accessing
of this node.

Test: manual test
Bug: 72196764
Change-Id: Iff283bf4f8dcf305785891d142e50afd6dc65875
---
 libwifi_hal/wifi_hal_common.cpp | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/libwifi_hal/wifi_hal_common.cpp b/libwifi_hal/wifi_hal_common.cpp
index 4bae629d0..aac65e1d4 100644
--- a/libwifi_hal/wifi_hal_common.cpp
+++ b/libwifi_hal/wifi_hal_common.cpp
@@ -99,8 +99,20 @@ int wifi_change_driver_state(const char *state) {
   int len;
   int fd;
   int ret = 0;
+  int count = 5; /* wait at most 1 second for completion */
 
   if (!state) return -1;
+
+  do {
+    if (access(WIFI_DRIVER_STATE_CTRL_PARAM, R_OK|W_OK) == 0)
+      break;
+    usleep(200000);
+  } while (--count > 0);
+    if (count == 0) {
+      PLOG(ERROR) << "Failed to access driver state control param " << strerror(errno) << ", " << errno;
+      return -1;
+  }
+
   fd = TEMP_FAILURE_RETRY(open(WIFI_DRIVER_STATE_CTRL_PARAM, O_WRONLY));
   if (fd < 0) {
     PLOG(ERROR) << "Failed to open driver state control param";
-- 
2.17.1

